### PR TITLE
SAK-47878 - Improve the configuration and launch experience for sakai.web.168 tool

### PR DESCRIPTION
**Pull request by csev**
Adrian - this makes a new web component for the sakai-lti-popup and adds some features to the sakai-lti-iframe code to cope with bad urls.   But there is a TODO that I could not make work because the then() clause does not finish before the constructor finishes :(

### CODE REVIEW

**Comment by csev** [comment]
So @adrianfish  I figured out the then() syntax and finished this up - it is ready for review.

File: webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js [hunk 1]
```diff
@@ -1,28 +1,76 @@
...
     const randomId = Math.floor(Math.random() * 1000000);
     this.randomId = randomId;
+    this.newWindowText = null;
+    this.loadTranslations("lti").then(t => {
+      this.i18n = t
+      if ( this.newWindowText == null ) this.newWindowText = this.i18n.new_window_text;
+    } );
+
+    setTimeout( (elem) => {
+      console.log('elem', elem);
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js:16]: I'd go for log.debug here.

File: webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js [hunk 2]
```diff
@@ -1,28 +1,76 @@
...
+      if ( typeof elem == 'undefined' ) return;
+      console.log("timeout check", elem.randomId, elem.launchUrl);
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js:18]: Again, I'd use log.debug, try and keep the normal console a bit less cluttered.

File: webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js [hunk 3]
```diff
@@ -1,28 +1,76 @@
...
+      // Only check off-server launches
+      if ( elem.launchUrl.indexOf('http://') != 0 && elem.launchUrl.indexOf('https://') != 0 ) return;
+
+      console.log('Actually checking...');
+
+      const myframe = $("#sakai-lti-iframe-" + elem.randomId);
+      const mybutton = $("#sakai-lti-button-" + elem.randomId);
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js:25]:
```suggestion
      const myframe = document.getElementById(`sakai-lti-iframe-${elem.randomId}`);
      const mybutton = document.getElementById(`sakai-lti-button-${elem.randomId}`);
```

File: webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js [hunk 4]
```diff
@@ -1,28 +1,76 @@
...
+      let loaded = false;
 
-    window.addEventListener('message', (e) => {
       try {
-        const idval = `sakai-lti-iframe-${randomId}`;
-        // https://stackoverflow.com/questions/15329710/postmessage-source-iframe
+        const iframeLoc = String(myframe.contentWindow.location);
+        console.debug('iframeLoc', iframeLoc);
+        const windowLoc = String(window.location);
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js:31]:
```suggestion
        const windowLoc = window.location.href;
```

File: webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js [hunk 5]
```diff
@@ -1,28 +1,76 @@
...
+    }, 3000, this);
+
+    window.addEventListener('message', e => {
+      try {
+        const idval = `sakai-lti-iframe-${randomId}`; // https://stackoverflow.com/questions/15329710/postmessage-source-iframe
+
         let frame_id = false;
         let allow_resize = false;
-        Array.prototype.forEach.call(document.getElementsByTagName('iframe'), (element) => {
+        Array.prototype.forEach.call(document.getElementsByTagName('iframe'), element => {
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-lti-iframe.js:65]:
```suggestion
        Array.from(document.getElementsByTagName('iframe')).forEach(element => {
```

File: webcomponents/tool/src/main/frontend/js/sakai-lti-popup.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  shouldUpdate() {
+    return this.preLaunchText && this.postLaunchText && this.launchUrl;
+  }
+
+  render() {
+    return html`
+<div>
+<button id="sakai-lti-popup-${this.randomId}" 
+    class="btn btn-primary" role="button"
+    onclick="
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-lti-popup.js:44]: You should use the @click handler. Just create a function in your class with all the stuff in the onclick handler, then add @click to the button tag. So, you'd add @click=${this.handlePopup} to the tag.

**Review by adrianfish — CHANGES_REQUESTED** [review]
A few changes needed, not much :)

**Comment by csev** [comment]
Hey @adrianfish - I think I got them all - thanks for the @click pointer.  Elegant.

**Review by adrianfish — APPROVED** [review]
GTG :)

**Pull request merged**